File: leet-code/739.每日温度.js
/*
 * @lc app=leetcode.cn id=739 lang=javascript
 *
 * [739] 每日温度
 */

// @lc code=start
/**
 * @param {number[]} temperatures
 * @return {number[]}
 */
var dailyTemperatures = function (temperatures) {
  // 利用单调栈
  const res = Array(temperatures.length).fill(0)
  const stack = []
  for (let i = 0; i < temperatures.length; i++) {
    const item = temperatures[i];
    while (stack.length) {
      const idx = stack.pop() // 栈顶
      if (item > temperatures[idx]) {
        // 大于,获取当前元素和栈顶的下标差
        res[idx] = i - idx
      } else {
        // 小于等于
        stack.push(idx)
        break
      }
    }
    stack.push(i)
  }
  return res
};
// @lc code=end

File: leet-code-rs/src/solutions/739_每日温度.rs
/*
 * @lc app=leetcode.cn id=739 lang=rust
 *
 * [739] 每日温度
 */

// @lc code=start

#[allow(unused)]
impl Solution {
    // 解题思路：单调递减栈
    pub fn daily_temperatures(temperatures: Vec<i32>) -> Vec<i32> {
        let mut single_stack = vec![];
        let mut res = vec![0; temperatures.len()];

        for (i, n) in temperatures.iter().enumerate() {
            while !single_stack.is_empty() && temperatures[*single_stack.last().unwrap()] < *n {
                let pre = single_stack.pop().unwrap();
                res[pre] = (i - pre) as i32;
            }

            single_stack.push(i);
        }

        // 剩下的是单调递减的温度，代表往后不会升温，设置值为 0，初始化即为 0，这里跳过
        // while !single_stack.is_empty() {
        //     let idx = single_stack.pop().unwrap();
        //     res[idx] = 0;
        // }

        return res;
    }
}
// @lc code=end

#[allow(unused)]
struct Solution;

#[cfg(test)]
mod test {
    #[allow(unused)]
    use super::*;

    #[test]
    fn test_daily_temperatures() {
        //println!("res is {:?}", Solution::daily_temperatures(vec![]));
        //assert_eq!(Solution::daily_temperatures(vec![]),[]);
        assert!(true)
    }
}
